sync/azuredatabricks.py

```diff
@@ -0,0 +1,411 @@
+                        message=f"Can list objects at {parsed_log_url.geturl()}",
+                    )
+                )
+        else:
+            report.append(
+                AccessReportLine(
+                    name="Log Access",
+                    status=AccessStatusCode.RED,
+                    message=f"scheme in {parsed_log_url.geturl()} is not supported",
+                )
+            )
+
+    return report
+
+
+def _get_cluster_report(
+    cluster_id: str, plan_type: str, compute_type: str, allow_incomplete: bool
+) -> Response[AzureDatabricksClusterReport]:
+    # Cluster `terminated_time` can be a few seconds after the start of the next task in which
+    # this may be executing.
+    cluster_response = _wait_for_cluster_termination(cluster_id, poll_seconds=5)
+    if cluster_response.error:
+        return cluster_response
+
+    cluster = cluster_response.result
+
+    instances = _get_cluster_instances(cluster)
+    if instances.error:
+        if allow_incomplete:
+            logger.warning(instances.error)
+        else:
+            return instances
+
+    cluster_events = _get_all_cluster_events(cluster_id)
+    return Response(
+        result=AzureDatabricksClusterReport(
+            plan_type=plan_type,
+            compute_type=compute_type,
+            cluster=cluster,
+            cluster_events=cluster_events,
+            instances=instances.result,
+        )
+    )
+
+
+sync._databricks._get_cluster_report = _get_cluster_report
+
+
+def _get_cluster_instances(cluster: dict) -> Response[dict]:
+    cluster_instances = None
+
+    cluster_log_dest = _cluster_log_destination(cluster)
+
+    if cluster_log_dest:
+        (_, filesystem, bucket, base_prefix) = cluster_log_dest
+
+        cluster_id = cluster["cluster_id"]
+        spark_context_id = cluster["spark_context_id"]
+        cluster_instances_file_key = (
+            f"{base_prefix}/sync_data/{spark_context_id}/cluster_instances.json"
+        )
+
+        cluster_instances_file_response = None
+        if filesystem == "dbfs":
+            cluster_instances_file_response = _get_cluster_instances_from_dbfs(
+                cluster_instances_file_key
+            )
+
+        cluster_instances = (
+            orjson.loads(cluster_instances_file_response)
+            if cluster_instances_file_response
+            else None
+        )
+
+    # If this cluster does not have the "Sync agent" configured, attempt a best-effort snapshot of the instances that
+    #  are associated with this cluster
+    if not cluster_instances:
+
+        resource_group_name = _get_databricks_resource_group_name()
+
+        compute = _get_azure_client(ComputeManagementClient)
+        if resource_group_name:
+            vms = compute.virtual_machines.list(resource_group_name=resource_group_name)
+        else:
+            logger.warning("Failed to find Databricks managed resource group")
+            vms = compute.virtual_machines.list_all()
```

Comment:
are these responses not paginated? The docs are a little hard to grok, but I see a `next` field in the example response here - https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/list?tabs=HTTP

Having to go through all VMs in a workspace seems like it will be slow... is there really no way to query based on tags?

Reply:
Yeah, I'm kinda concerned about that too. After some searching I concluded that you really can't filter by tags 🤦

I think the paging goes on under the hood as you iterate through the result - I didn't come across any paging parameters, but I noted that the request doesn't happen until you start pulling vms out of `vms`.

Reply:
> After some searching I concluded that you really can't filter by tags 🤦

Oof. Something to watch out for, for sure

> I think the paging goes on under the hood as you iterate through the result

Ah, very possible! Yeah, I was looking at the REST reference, not the Python SDK (docs on the Python SDK seem quite hard to find??)